plugins/com.google.cloud.tools.eclipse.usagetracker/src/com/google/cloud/tools/eclipse/usagetracker/AnalyticsPingManager.java

```diff
@@ -102,11 +103,27 @@ static void registerOptInStatus(boolean optedIn) {
     }
   }
 
+  /**
+   * Sends a usage metric to Google Analytics.
+   * <p/>
+   * If the user has never seen the opt-in dialog or set the opt-in preference beforehand,
+   * this method can potentially present an opt-in dialog at the top workbench level. If you
+   * are calling this method inside another modal dialog, consider using {@link #sendPing(
+   * String, String, String, Shell)} and pass the {@Shell} of currently open modal dialog.
+   * (Otherwise, the opt-in dialog won't be able to get input until the workbench can get input.)
+   */
   public static void sendPing(String eventName, String metadataKey, String metadataValue) {
+    sendPing(eventName, metadataKey, metadataValue, null);
+  }
+
+  public static void sendPing(String eventName, String metadataKey, String metadataValue,
+      Shell parentShell) {
     if (Platform.inDevelopmentMode() || !isTrackingIdDefined() || !hasUserOptedIn()) {
       return;
     }
 
+    showOptInDialog(parentShell);
```

Comment:
This code implies the sendPing will block…  Rather than pass a shell, I wonder if we should instead have a flag to say _it's ok to query the user_ vs _drop event if user hasn't yet agreed_.